### DISCUSSION TITLE
__THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__()

### DISCUSSION THREAD

**Discussion by camierjs** [GLVis/glvis#141] [Q&A] [answered]
I'm hitting this issue on `macOS Catalina 10.15.7`:
```
The process has forked and you cannot use this CoreFoundation functionality safely. You MUST exec().
Break on __THE_PROCESS_HAS_FORKED_AND_YOU_CANNOT_USE_THIS_COREFOUNDATION_FUNCTIONALITY___YOU_MUST_EXEC__() to debug
```
Has anyone seen/hit this before?
This shows up after adding `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` in the shell to avoid a first 
```[NSApplication initialize] may have been in progress in another thread when fork() was called.``` error.

**Comment by tzanio** [comment]
@kanye-quest and @jandrej can you help with that?

**Comment by tzanio** [comment]
Quoting from https://glvis.org/options-and-use/#server-mode:

> On legacy Mac machines with OS X Leopard, the server needs to be started with
>
> `glvis -mac`
>
> This is due to the fact that Mac OS X returns an error when fork() is called without an immediate exec(). Note that this option is not necessary on newer versions of OS X.

@camierjs, can you try that? I really hope they haven't reintroduced this issue in the new versions of macOS 😞

**Comment by camierjs** [comment]
It seems to work better - but I'm still getting hangs with multiple sockets.

Maybe it's related to the `glvis-saved.*` files?

**Comment by camierjs** [comment]
Reverted to `Command_Line_Tools_for_Xcode_12.2` and `Xcode_12.2` runtime libraries w/o better success.

Removing the `multi_session` mode allows to at least get one window working properly.

**Comment by tzanio** [comment]
I can't reproduce this and most Google searches (e.g. "catalina corefoundation fork exec") return result related to Python.

Is it possible that you updated other things besides Xcode? Can you confirm that you see this error with simple `./glvis -m ../mfem/data/star.mesh`?

**Comment by camierjs** [comment]
It turns out that the SDL version [2.0.14](https://www.libsdl.org/release/SDL2-2.0.14.tar.gz) introduces this behavior - [2.0.12](https://www.libsdl.org/release/SDL2-2.0.12.tar.gz) works fine.

**Comment by tzanio** [comment]
@kanye-quest, heads up that we may run into this problem ☝️ when updating SDL2 on Mac

**Comment by camierjs** [comment, marked as answer]
I've extracted the `sdl2.rb` that can install the `2.0.12` SDL2 library through brew with `brew install ./sdl2.rb`

[sdl2.rb.txt](https://github.com/GLVis/glvis/files/5788895/sdl2.rb.txt)

**Reply by zahurullah123** [reply]
@camierjs Thanks for the script, it also sovled my probem.

**Comment by tzanio** [comment]
This issue will be addressed by the work in https://github.com/GLVis/glvis/pull/149

**Comment by tzanio** [comment]
If the https://github.com/GLVis/glvis/discussions/141#discussioncomment-269943  solution doesn't work for you, you can build SDL2 manually, as described in https://github.com/GLVis/glvis/blob/master/INSTALL#L209-L223.

Here, for completeness, are the necessary commands (on a Mac, from the directory containing glvis/):

```
wget "https://www.libsdl.org/release/SDL2-2.0.12.tar.gz"
tar zxf SDL2-2.0.12.tar.gz
cd SDL2-2.0.12
./configure --prefix=`dirname $PWD`/SDL2 --disable-video-x11
make -j
make install
cd ..
```

**Comment by tzanio** [comment]
This should be resolved in `master` as of https://github.com/GLVis/glvis/commit/b513773a803dd99117c4ef8348465f3ae5674d86 (Jul 9th, 2021)
